src/github.com/getlantern/peerscanner/web.go

```diff
@@ -0,0 +1,150 @@
+	}
+
+	log.Debug("About to serve")
+	err = http.Serve(l, nil)
+	if err != nil {
+		log.Fatalf("Unable to serve: %s", err)
+	}
+}
+
+// register is the entry point for peers registering themselves with the service.
+// If peers are successfully vetted, they'll be added to the DNS round robin.
+func register(resp http.ResponseWriter, req *http.Request) {
+	name, ip, port, err := getHostInfo(req)
+	if err == nil && port != 443 {
+		err = fmt.Errorf("Port %d not supported, only port 443 is supported", port)
+	}
+	if err != nil {
+		resp.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintln(resp, err.Error())
+		return
+	}
+
+	h := getOrCreateHost(name, ip)
+	online, connectionRefused, timedOut := h.status()
+	if timedOut {
+		log.Debugf("%v timed out waiting for status, returning 500 error", h)
+		resp.WriteHeader(500)
+		fmt.Fprintf(resp, "Timed out waiting for status")
+		return
+	}
+
+	if online {
+		resp.WriteHeader(200)
+		fmt.Fprintln(resp, "Connectivity to proxy confirmed")
+		return
+	}
+
+	// Note this may not work across platforms, but the intent
+	// is to tell the client if the connection was flat out
+	// refused as opposed to timed out in order to allow them
+	// to configure their router if possible.
+	if connectionRefused {
+		// 417 response code.
+		resp.WriteHeader(http.StatusExpectationFailed)
+		fmt.Fprintln(resp, "No connectivity to proxy - connection refused")
+	} else {
+		// 408 response code.
+		resp.WriteHeader(http.StatusRequestTimeout)
+		fmt.Fprintln(resp, "No connectivity to proxy - test request timed out")
+	}
+}
+
+// unregister is the HTTP endpoint for removing peers from DNS. Peers are
+// unregistered based on their ip (not their name).
+func unregister(resp http.ResponseWriter, req *http.Request) {
+	_, ip, _, err := getHostInfo(req)
+	if err != nil {
+		resp.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintln(resp, err.Error())
+		return
+	}
+
+	h := getHostByIp(ip)
+	msg := "Host not registered"
+	if h != nil {
+		h.unregister()
+		msg = "Host unregistered"
+	}
+	resp.WriteHeader(200)
+	fmt.Fprintln(resp, msg)
+}
+
+func getHostInfo(req *http.Request) (name string, ip string, port int, err error) {
+	name = req.FormValue("name")
+	if name == "" {
+		err = fmt.Errorf("Please specify a name")
+		return
+	}
+	ip = clientIpFor(req, name)
+	if ip == "" {
+		err = fmt.Errorf("Unable to determine IP address")
+		return
+	}
+	portString := req.FormValue("port")
+
+	if portString != "" {
+		port, err = strconv.Atoi(portString)
+		if err != nil {
+			err = fmt.Errorf("Received invalid port for %v - %v: %v", name, ip, portString)
+		}
+	}
+
+	return
+}
+
+func clientIpFor(req *http.Request, name string) string {
+	// Client requested their info
+	clientIp := req.Header.Get("X-Peerscanner-Forwarded-For")
+	if clientIp == "" {
+		clientIp = req.Header.Get("X-Forwarded-For")
+	}
+	if clientIp == "" && isFallback(name) {
+		// Use direct IP for fallbacks
+		clientIp = strings.Split(req.RemoteAddr, ":")[0]
```

Comment:
If we put peerscanner behind a proxy (such as nginx), this `RemoteAddr` would be the IP of the proxy rather than the IP of the remote machine. This could be solved by accepting `X-Forwarded-For` or `X-Peerscanner-Forwarded-For` headers as above, but how can we trust or test those values?

The nginx web server has a (realip_module](http://nginx.org/en/docs/http/ngx_http_realip_module.html) that also trusts `X-Forwarded-For` headers, but only from a fixed whitelist. It don't think we could implement a whitelist since we don't know the servers that are going to try to register with peerscanner but maybe we could test them somehow before trusting the IP we're receiving.